File: backend/src/test/java/com/back/back9/domain/coin/controller/CoinControllerTest.java
package com.back.back9.domain.coin.controller;

import com.back.back9.domain.coin.entity.Coin;
import com.back.back9.domain.coin.repository.CoinRepository;
import com.back.back9.domain.coin.service.CoinService;
import com.back.back9.global.error.ErrorException;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.DisplayName;
import org.junit.jupiter.api.Tag;
import org.junit.jupiter.api.Test;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc;
import org.springframework.boot.test.context.SpringBootTest;
import org.springframework.http.MediaType;
import org.springframework.test.context.ActiveProfiles;
import org.springframework.test.web.servlet.MockMvc;
import org.springframework.test.web.servlet.ResultActions;
import org.springframework.transaction.annotation.Transactional;

import java.util.List;

import static org.junit.jupiter.api.Assertions.assertThrows;
import static org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*;
import static org.springframework.test.web.servlet.result.MockMvcResultHandlers.print;
import static org.springframework.test.web.servlet.result.MockMvcResultMatchers.*;

@SpringBootTest
@ActiveProfiles("test")
@AutoConfigureMockMvc
@Transactional
@Tag("coin")
public class CoinControllerTest {
    @Autowired
    private MockMvc mvc;

    @Autowired
    private CoinService coinService;

    @Autowired
    private CoinRepository coinRepository;

    private Coin coin1;
    private Coin coin2;
    private Coin coin3;
    private Coin coin4;

    @BeforeEach
    void setUp() {
        coinRepository.deleteAll();

        coin1 = coinRepository.save(new Coin("BTC","비트코안","Bitcoin"));
        coin2 = coinRepository.save(new Coin("ETH", "이더리움", "Ethereum"));
        coin3 = coinRepository.save(new Coin("XRP","리플","Ripple"));
        coin4 = coinRepository.save(new Coin("DOGE","도지코인","Dogecoin"));
    }


    @Test
    @DisplayName("Coin 전체 조회")
    void getCoins() throws Exception {
        ResultActions resultActions = mvc
                .perform(
                        get("/api/v1/coins")
                )
                .andDo(print());

        List<Coin> testCoin = coinService.findAll();

        resultActions
                .andExpect(handler().handlerType(CoinController.class))
                .andExpect(handler().methodName("getCoins"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.length()").value(4));

        for (int i = 0; i < testCoin.size(); i++) {
            Coin coin = testCoin.get(i);
            resultActions
                    .andExpect(jsonPath("$[%d].id".formatted(i)).value(coin.getId()))
                    .andExpect(jsonPath("$[%d].symbol".formatted(i)).value(coin.getSymbol()))
                    .andExpect(jsonPath("$[%d].koreanName".formatted(i)).value(coin.getKoreanName()))
                    .andExpect(jsonPath("$[%d].englishName".formatted(i)).value(coin.getEnglishName()));
        }

    }

    @Test
    @DisplayName("Coin 단건 조회")
    void getCoin() throws Exception {
        ResultActions resultActions = mvc
                .perform(
                        get("/api/v1/coins/" + coin1.getId())
                )
                .andDo(print());

        resultActions
                .andExpect(handler().handlerType(CoinController.class))
                .andExpect(handler().methodName("getCoin"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.symbol").value(coin1.getSymbol()));
    }

    @Test
    @DisplayName("Coin 단건 조회, 404")
    void getCoin2() throws Exception {
        int id = 999;

        ResultActions resultActions = mvc
                .perform(
                        get("/api/v1/coins/" + id)
                )
                .andDo(print());

        resultActions
                .andExpect(handler().handlerType(CoinController.class))
                .andExpect(handler().methodName("getCoin"))
                .andExpect(status().isNotFound())
                .andExpect(jsonPath("$.status").value("404"));
    }

    @Test
    @DisplayName("Coin 추가")
    void addCoin() throws Exception {
        ResultActions resultActions = mvc
                .perform(
                        post("/api/v1/coins")
                                .contentType(MediaType.APPLICATION_JSON)
                                .content("""
                                        {
                                        "symbol" : "BTC new",
                                        "koreanName" : "비트코인 new",
                                        "englishName" : "Bitcoint new"
                                        }
                                        """)
                ).andDo(print());

        Coin coin = coinService.findLastest().get();


        resultActions
                .andExpect(handler().handlerType(CoinController.class))
                .andExpect(handler().methodName("addCoin"))
                .andExpect(status().isOk())
                .andExpect(jsonPath("$.id").value(coin.getId()))
                .andExpect(jsonPath("$.symbol").value(coin.getSymbol()))
                .andExpect(jsonPath("$.koreanName").value(coin.getKoreanName()))
                .andExpect(jsonPath("$.englishName").value(coin.getEnglishName()));
    }

    @Test
    @DisplayName("코인 추가, Without symbol")
    void addCoin2() throws Exception {
        ResultActions resultActions = mvc
                .perform(
                        post("/api/v1/coins")
                                .contentType(MediaType.APPLICATION_JSON)
                                .content("""
                                        {
                                        "symbol" : "",
                                        "koreanName" : "비트코인 new",
                                        "englishName" : "Bitcoint new"
                                        }
                                        """)
                ).andDo(print());

        resultActions
                .andExpect(handler().handlerType(CoinController.class))
                .andExpect(handler().methodName("addCoin"))
                .andExpect(status().isBadRequest())
                .andExpect(jsonPath("$.status").value("400"));
    }

    @Test
    @DisplayName("Coin 삭제")
    void deleteCoin() throws Exception {
        ResultActions resultActions = mvc
                .perform(
                        delete("/api/v1/coins/" + coin1.getId())
                )
                .andDo(print());

        resultActions
                .andExpect(handler().handlerType(CoinController.class))
                .andExpect(handler().methodName("deleteCoin"))
                .andExpect(status().isOk());

        assertThrows(ErrorException.class, () -> {
            coinService.findById(coin1.getId());
        });
    }

    @Test
    @DisplayName("Coin 수정")
    void modifyCoin() throws Exception {
        ResultActions resultActions = mvc
                .perform(
                        put("/api/v1/coins/" + coin1.getId())
                                .contentType(MediaType.APPLICATION_JSON)
                                .content("""
                                        {
                                        "symbol" : "BTC 수정",
                                        "koreanName" : "비트코인 수정",
                                        "englishName" : "Bitcoint 수정"
                                        }
                                        """)
                ).andDo(print());

        Coin coin = coinService.findById(coin1.getId());

        resultActions
                .andExpect(handler().handlerType(CoinController.class))
                .andExpect(handler().methodName("modifyCoin"))
                .andExpect(status().isOk());
    }
}

File: backend/src/test/kotlin/com/back/back9/domain/coin/controller/CoinControllerTest.kt
package com.back.back9.domain.coin.controller

import com.back.back9.domain.coin.entity.Coin
import com.back.back9.domain.coin.repository.CoinRepository
import com.fasterxml.jackson.databind.ObjectMapper
import org.junit.jupiter.api.BeforeEach
import org.junit.jupiter.api.DisplayName
import org.junit.jupiter.api.Tag
import org.junit.jupiter.api.Test
import org.springframework.beans.factory.annotation.Autowired
import org.springframework.boot.test.autoconfigure.web.servlet.AutoConfigureMockMvc
import org.springframework.boot.test.context.SpringBootTest
import org.springframework.http.MediaType
import org.springframework.test.context.ActiveProfiles
import org.springframework.test.web.servlet.MockMvc
import org.springframework.test.web.servlet.request.MockMvcRequestBuilders.*
import org.springframework.test.web.servlet.result.MockMvcResultHandlers.print
import org.springframework.test.web.servlet.result.MockMvcResultMatchers.*
import org.springframework.transaction.annotation.Transactional

@SpringBootTest
@ActiveProfiles("test")
@AutoConfigureMockMvc(addFilters = false)
@Transactional
@Tag("coin")
class CoinControllerTest @Autowired constructor(
    private val mvc: MockMvc,
    private val coinRepository: CoinRepository,
    private val objectMapper: ObjectMapper // JSON 파싱을 위해 추가
) {

    private lateinit var coin1: Coin

    @BeforeEach
    fun setUp() {
        coinRepository.deleteAll()
        // 테스트 데이터는 필요한 만큼만 생성합니다.
        coin1 = coinRepository.save(Coin(symbol = "BTC", koreanName = "비트코인", englishName = "Bitcoin"))
        coinRepository.save(Coin(symbol = "ETH", koreanName = "이더리움", englishName = "Ethereum"))
    }

    @Test
    @DisplayName("Coin 전체 조회")
    fun getCoins() {
        mvc.perform(get("/api/v1/adm/coins"))
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(CoinController::class.java))
            .andExpect(handler().methodName("getCoins"))
            .andExpect(jsonPath("$.length()").value(2))
            .andExpect(jsonPath("$[0].symbol").value("BTC"))
            .andExpect(jsonPath("$[1].symbol").value("ETH"))
            .andDo(print())
    }

    @Test
    @DisplayName("Coin 단건 조회")
    fun getCoin() {
        mvc.perform(get("/api/v1/adm/coins/${coin1.id}"))
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(CoinController::class.java))
            .andExpect(handler().methodName("getCoin"))
            .andExpect(jsonPath("$.id").value(coin1.id!!))
            .andExpect(jsonPath("$.symbol").value(coin1.symbol))
            .andDo(print())
    }

    @Test
    @DisplayName("Coin 단건 조회, 404")
    fun getCoinNotFound() {
        mvc.perform(get("/api/v1/adm/coins/99999"))
            .andExpect(status().isNotFound)
            .andExpect(handler().handlerType(CoinController::class.java))
            .andExpect(handler().methodName("getCoin"))
            .andExpect(jsonPath("$.status").value("404"))
            .andDo(print())
    }

    @Test
    @DisplayName("Coin 추가")
    fun addCoin() {
        val newCoinInfo = mapOf(
            "symbol" to "XRP",
            "koreanName" to "리플",
            "englishName" to "Ripple"
        )

        mvc.perform(
            post("/api/v1/adm/coins")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(newCoinInfo))
        )
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(CoinController::class.java))
            .andExpect(handler().methodName("addCoin"))
            // 응답값 자체를 검증
            .andExpect(jsonPath("$.id").exists()) // ID가 생성되었는지 확인
            .andExpect(jsonPath("$.symbol").value("XRP"))
            .andExpect(jsonPath("$.koreanName").value("리플"))
            .andDo(print())
    }

    @Test
    @DisplayName("코인 추가, symbol 누락 시 400 Bad Request")
    fun addCoinWithoutSymbol() {
        val newCoinInfo = mapOf(
            "symbol" to "",
            "koreanName" to "비트코인 new",
            "englishName" to "Bitcoin new"
        )

        mvc.perform(
            post("/api/v1/adm/coins")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(newCoinInfo))
        )
            .andExpect(status().isBadRequest)
            .andExpect(handler().handlerType(CoinController::class.java))
            .andExpect(handler().methodName("addCoin"))
            .andExpect(jsonPath("$.status").value("400"))
            .andDo(print())
    }

    @Test
    @DisplayName("Coin 삭제")
    fun deleteCoin() {
        // 1. 삭제 API 호출
        mvc.perform(delete("/api/v1/adm/coins/${coin1.id}"))
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(CoinController::class.java))
            .andExpect(handler().methodName("deleteCoin"))
            .andDo(print())

        // 2. 정말 삭제되었는지 확인하기 위해 다시 조회 API를 호출
        mvc.perform(get("/api/v1/adm/coins/${coin1.id}"))
            .andExpect(status().isNotFound) // 404 Not Found 응답을 기대
            .andDo(print())
    }

    @Test
    @DisplayName("Coin 수정")
    fun modifyCoin() {
        val modifiedInfo = mapOf(
            "symbol" to "BTC-MODIFIED",
            "koreanName" to "비트코인(수정)",
            "englishName" to "Bitcoin-Modified"
        )

        // 1. 수정 API 호출
        mvc.perform(
            put("/api/v1/adm/coins/${coin1.id}")
                .contentType(MediaType.APPLICATION_JSON)
                .content(objectMapper.writeValueAsString(modifiedInfo))
        )
            .andExpect(status().isOk)
            .andExpect(handler().handlerType(CoinController::class.java))
            .andExpect(handler().methodName("modifyCoin"))
            .andDo(print())

        // 2. 정말 수정되었는지 확인하기 위해 다시 조회 API를 호출
        mvc.perform(get("/api/v1/adm/coins/${coin1.id}"))
            .andExpect(status().isOk)
            .andExpect(jsonPath("$.symbol").value("BTC-MODIFIED"))
            .andExpect(jsonPath("$.koreanName").value("비트코인(수정)"))
            .andDo(print())
    }
}
